### ISSUE TITLE
Add a way to undo a specific change just before saving?

### ISSUE THREAD

**Issue by helmo** [openstreetmap/iD#3445]
When working in iD you get a list of changes below the 'Changeset comment' field when you try to save.

Is there a way to undo a deletion without using the undo button?

Example: 
1. deleted address,
2. create poi,
3. modify street,
42. more ...

When I click a 'create' of 'modify' change that change is highlighted, letting me correct it (if you still know what changed). But the deletion... there is nothing to highlight.

A little 'x' in the list of changes to remove that one change would be a solution for me.

I asked this a few days ago on the [help site](https://help.openstreetmap.org/questions/52097/is-there-a-way-to-undo-a-specific-change-just-before-saving) but was directed here.

**Comment by bhousel** [comment]
There are a few cases where this is easy to do, and a lot of cases where this is difficult or impossible to do. I'll think about more about this issue, but my gut is telling me :no_good_man:

**Comment by jfirebaugh** [comment]
Previously: #537.

**Comment by bhousel** [comment]
Yes, I was also thinking about exposing the revert action for this.  But need to wrap my head around the limitations of where it's safe to use.

**Comment by EvanCarroll** [comment]
I just edited the map for 2 hours. Upon saving I realized I must have made a mistake and moved a road. I wish doing this would immediately trigger a warning. Not being able to select that change and remove it is devastating. I just added a whole apartment complex.

**Comment by 1ec5** [comment]
> Not being able to select that change and remove it is devastating.

Oof, I know how this feels. Hopefully it wasn’t a problem to move the road back. But in the future, if you run into a situation where you _really_ need to save your hard work but accidentally made an unrelated change, you can export your changes and selectively save them outside of iD:

1. Click Save, then click “Download osmChange file” at the bottom of the panel.
2. Open the .osc file in JOSM or [Level0](https://level0.osmz.ru/). In Level0, choose the file, then click “Add to editor”.
3. Selectively remove the edits you don’t want to save. Level0 has [a special syntax](https://wiki.openstreetmap.org/wiki/Level0L) that’s easier than the raw XML in the .osc file. Remove the entries for the nodes you accidentally moved, based on their node IDs. You may also find the affected way in the file, but as long as you didn’t make any other accidental changes, you can leave that way’s entry untouched.
4. Upload your changes.

**Comment by EvanCarroll** [comment]
This is what I did, but ideally, this functionality would be in the web app and users wouldn't have to download Java and another tool. Thankfully this method worked out great.

**Comment by Kogacarlo** [comment]
Where is the .osc file of my last changeset? I want to undo one thing.

**Comment by 1ec5** [comment]
The option to download an .osc only exists until you upload the changeset. After that, if you want to undo anything, you need to undo it manually or use a tool like JOSM’s reverter plugin or [RevertUI](http://revert.osmz.ru/).

**Comment by Kogacarlo** [comment]
Thanks. By using the reverter plugin the old nodes were visible in a _second layer_. I moved some nodes back to their old position, deleted the _second layer_ and uploaded. It worked well :-)

**Comment by aandrejevas** [comment]
please, can this be implemented?

**Comment by aandrejevas** [comment]
As a workaround, using the ID editor it is now possible to upload osmChange file (https://github.com/openstreetmap/iD/pull/10407). So it is now simpler to undo changes:

1. In the ID editor click save, then click "Download osmChange file" at the bottom of the panel.
2. I opened the downloaded file with vscode, specified that the language is html (I did not have a xml formatter extension installed) and then formatted the file. After the file is formatted it is really easy to see what nodes you need to delete. So delete those nodes and save the file. Or use some other method to modify the .osc file.
4. In the ID editor discard previous changes and press "Map Data", there you can find an option to "Load edits from file". Load the saved file and then save and upload the changeset.

**Comment by matkoniecz** [comment]
That is basically re-implementing git inside iD (cherry picking, rebasing etc), but with requirement to make it easy to use and explain.

That is horrific amount of work, and will generate endless support requests.

**Comment by tyrasd** [comment]
Closing as out of scope / in favour of #537. PS: with #10407 it would be possible to do the manual workaround by manipulating the osm-change file.
